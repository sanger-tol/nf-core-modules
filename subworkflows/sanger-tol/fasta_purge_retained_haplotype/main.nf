/*
Adapted from
https://github.com/NBISweden/Earth-Biogenome-Project-pilot/blob/5ec2002638055bb8396857a8ee418bf86188fc59/subworkflows/purge_dups/main.nf
*/

/*
 * Workflow based around the purge_dups tool
 * https://github.com/dfguan/purge_dups
 */

include { CAT_CAT as CAT_PURGED_HAPS_TO_ALT         } from '../../../modules/nf-core/cat/cat'
include { MINIMAP2_ALIGN as MINIMAP2_ALIGN_ASSEMBLY } from '../../../modules/nf-core/minimap2/align'
include { PURGEDUPS_CALCUTS                         } from '../../../modules/nf-core/purgedups/calcuts'
include { PURGEDUPS_GETSEQS                         } from '../../../modules/nf-core/purgedups/getseqs'
include { PURGEDUPS_PBCSTAT                         } from '../../../modules/nf-core/purgedups/pbcstat'
include { PURGEDUPS_PURGEDUPS                       } from '../../../modules/nf-core/purgedups/purgedups'
include { PURGEDUPS_SPLITFA                         } from '../../../modules/nf-core/purgedups/splitfa'

include { FASTX_MAP_LONG_READS                      } from '../fastx_map_long_reads/main'

workflow PURGING {

    take:
    ch_assemblies             // [meta, hap1, hap2]
    ch_long_reads             // [meta, [reads]] - should be a value channel
    val_fastx_reads_per_chunk // integer: number of reads per chunk to map

    main:
    ch_versions = Channel.empty()

    //
    // Logic: split assemblies into primary and alternate
    //
    ch_assemblies_split = ch_assemblies
        | multiMap { meta, hap1, hap2 ->
            primary:  [meta, hap1]
            alternate:[meta, hap2]
        }

    //
    // Subworkflow: Map HiFi reads to primary contigs
    //
    FASTX_MAP_LONG_READS(
        ch_assemblies_split.primary,
        ch_long_reads,
        val_fastx_reads_per_chunk,
        false
    )
    ch_versions = ch_versions.mix(FASTX_MAP_LONG_READS.out.versions)

    //
    // Module: Create read depth histogram
    //
    PURGEDUPS_PBCSTAT(FASTX_MAP_LONG_READS.out.paf)
    ch_versions = ch_versions.mix(PURGEDUPS_PBCSTAT.out.versions)

    //
    // Module: Generate cutoffs based on histogram and kmer coverage
    //
    PURGEDUPS_CALCUTS(PURGEDUPS_PBCSTAT.out.stat)
    ch_versions = ch_versions.mix(PURGEDUPS_CALCUTS.out.versions)

    //
    // Module: Split assembly
    //
    PURGEDUPS_SPLITFA(ch_assemblies_split.primary)
    ch_versions = ch_versions.mix(PURGEDUPS_SPLITFA.out.versions)

    //
    // MODULE: PEFORM SELF ALIGNMENT
    //
    MINIMAP2_ALIGN_ASSEMBLY (
        PURGEDUPS_SPLITFA.out.split_fasta,
        [[:], []], // Trigger read to read alignment
        false,     // bam output
        false,     // bam index extension
        false,     // cigar in paf file
        false      // cigar in bam file
    )
    ch_versions = ch_versions.mix(MINIMAP2_ALIGN_ASSEMBLY.out.versions)

    //
    // Module: Purge haplotigs from primary assembly
    //
    ch_purgedups_input = PURGEDUPS_PBCSTAT.out.basecov
        | join(PURGEDUPS_CALCUTS.out.cutoff)
        | join(MINIMAP2_ALIGN_ASSEMBLY.out.paf)

    PURGEDUPS_PURGEDUPS(ch_purgedups_input)
    ch_versions = ch_versions.mix(PURGEDUPS_PURGEDUPS.out.versions)

    //
    // Module: Generate the primary and alternative contigs
    //
    ch_getseqs_input = ch_assemblies_split.primary
        | join(PURGEDUPS_PURGEDUPS.out.bed)

    PURGEDUPS_GETSEQS(ch_getseqs_input)
    ch_versions = ch_versions.mix(PURGEDUPS_GETSEQS.out.versions)

    //
    // Module: Combine the haplotigs purged from the primary back
    //         into the alternate assembly
    //
    ch_alt_to_cat = PURGEDUPS_GETSEQS.out.haplotigs
        | join(ch_assemblies_split.alternate)
        | map { meta, haps, alt -> [meta, [alt, haps]] }

    CAT_PURGED_HAPS_TO_ALT(ch_alt_to_cat)
    ch_versions = ch_versions.mix(CAT_PURGED_HAPS_TO_ALT.out.versions)

    //
    // Logic: Combine purged primary and alternate assemblies back into
    //        a single output channel
    //
    ch_assemblies  = PURGEDUPS_GETSEQS.out.purged
        | join(CAT_PURGED_HAPS_TO_ALT.out.file_out)

    emit:
    assemblies = ch_assemblies
    versions = ch_versions
}
