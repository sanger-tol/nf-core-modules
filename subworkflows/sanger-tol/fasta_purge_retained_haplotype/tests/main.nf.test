nextflow_workflow {

    name "Test Subworkflow FASTA_PURGE_RETAINED_HAPLOTYPE"
    script "../main.nf"
    workflow "FASTA_PURGE_RETAINED_HAPLOTYPE"

    tag "subworkflows"
    tag "subworkflows_sangertol"
    tag "subworkflows/fasta_purge_retained_haplotype"
    tag "subworkflows/fastx_map_long_reads"
    tag "subworkflows/../../modules/nf-core/minimap2/align"
    tag "subworkflows/../../modules/nf-core/purgedups/pbcstat"
    tag "subworkflows/../../modules/nf-core/purgedups/splitfa"
    tag "subworkflows/../../modules/nf-core/purgedups/getseqs"
    tag "subworkflows/../../modules/nf-core/purgedups/purgedups"
    tag "subworkflows/../../modules/nf-core/purgedups/calcuts"
    tag "subworkflows/../../modules/nf-core/cat/cat"

    config "./nextflow.config"

    setup {
        nfcoreInitialise("${launchDir}/library/")
        nfcoreInstall(
            "${launchDir}/library/",
            [
                "cat/cat",
                "minimap2/align",
                "minimap2/index",
                "samtools/merge",
                "samtools/markdup",
                "samtools/faidx",
                "purgedups/calcuts",
                "purgedups/getseqs",
                "purgedups/histplot",
                "purgedups/pbcstat",
                "purgedups/purgedups",
                "purgedups/splitfa"
            ]
        )
        nfcoreLink("${launchDir}/library/", "${baseDir}/modules/")
    }

    test("Undibacterium unclassified - hap1/hap2 - hifi") {

        when {

            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id: 'test' ], // meta map
                        file(params.modules_testdata_base_path + 'Undibacterium_unclassified/assembly/draft/baUndUnlc1.hic.hap1.p_ctg.fa.gz', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'Undibacterium_unclassified/assembly/draft/baUndUnlc1.hic.hap2.p_ctg.fa.gz', checkIfExists: true),
                    ]
                )
                input[1] = Channel.of(
                    [
                        [ id: 'test' ],
                        file(params.modules_testdata_base_path + 'Undibacterium_unclassified/genomic_data/baUndUnlc1/pacbio/fasta/HiFi.reads.fasta', checkIfExists: true),
                    ]
                )
                input[2] = 200
                """
            }
        }

        then {
            assert workflow.success
            assertAll(
                { assert snapshot(
                    workflow.out
                ).match() }
            )
        }

        cleanup {
            nfcoreUnlink("${launchDir}/library/", "${baseDir}/modules/nf-core")
        }

    }

    test("Undibacterium unclassified - hap1 only - hifi") {

        when {

            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id: 'test' ], // meta map
                        file(params.modules_testdata_base_path + 'Undibacterium_unclassified/assembly/draft/baUndUnlc1.hic.hap1.p_ctg.fa.gz', checkIfExists: true),
                        [],
                    ]
                )
                input[1] = Channel.of(
                    [
                        [ id: 'test' ],
                        file(params.modules_testdata_base_path + 'Undibacterium_unclassified/genomic_data/baUndUnlc1/pacbio/fasta/HiFi.reads.fasta', checkIfExists: true),
                    ]
                )
                input[2] = 200
                """
            }
        }

        then {
            assert workflow.success
            assertAll(
                { assert snapshot(
                    workflow.out
                ).match() }
            )
        }

        cleanup {
            nfcoreUnlink("${launchDir}/library/", "${baseDir}/modules/nf-core")
        }

    }

}
