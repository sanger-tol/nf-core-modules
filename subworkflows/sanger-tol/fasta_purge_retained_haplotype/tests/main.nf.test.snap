{
    "Undibacterium unclassified - hap1/hap2 - hifi": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.purged.fa:md5,f39e71df41f2d5eb58110e5222984dfa",
                        "test.fa:md5,cd38bf45ddba78787705257b840f6568"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test.hap.fa:md5,b89d4488c2451413b065b04bb6e08b64"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test.PB.stat:md5,18e18919c3875b0ef14d557d050543ae"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test"
                        },
                        "test.PB.base.cov:md5,fe7e7762231da2b809b11eebdfdf8ee3"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cutoffs:md5,7a0c3b9371c4d1505452f5436f173e1e"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test"
                        },
                        "test.calcuts.log:md5,bb16374f3a2bc6ec445b9ac0b584d05b"
                    ]
                ],
                "6": [
                    [
                        {
                            "id": "test"
                        },
                        "test.dups.bed:md5,dfb067e8acdfde5c6dbef54a2ce1e247"
                    ]
                ],
                "7": [
                    [
                        {
                            "id": "test"
                        },
                        "test.purge_dups.log:md5,f53e9e39a59a0e7870dd6e7aa047cf22"
                    ]
                ],
                "8": [
                    "versions.yml:md5,6538edf27c63c29d6942c5b0a1c31c3c",
                    "versions.yml:md5,65f957bbaf722197435a5107c6c6b128",
                    "versions.yml:md5,77c810fac8a6f2edaaf050d96f063d99",
                    "versions.yml:md5,a68e25631af50931cf951142081c8bd7",
                    "versions.yml:md5,ab1b6b619f2ad229f2b03e05fad2c409",
                    "versions.yml:md5,ab1b6b619f2ad229f2b03e05fad2c409",
                    "versions.yml:md5,ab1b6b619f2ad229f2b03e05fad2c409",
                    "versions.yml:md5,ab1b6b619f2ad229f2b03e05fad2c409",
                    "versions.yml:md5,ab1b6b619f2ad229f2b03e05fad2c409",
                    "versions.yml:md5,bceaceffe8da91afbb66e351b6a9678d",
                    "versions.yml:md5,c9411452b63f327d2a5b66a82b8fd621",
                    "versions.yml:md5,d9e6066e723f9a40b1e22c33e37f2edf",
                    "versions.yml:md5,f7b0f31fa50eff25352b3aac997dfe71",
                    "versions.yml:md5,f8a504bd74283e317e5e474345c3badc"
                ],
                "purged_assemblies": [
                    [
                        {
                            "id": "test"
                        },
                        "test.purged.fa:md5,f39e71df41f2d5eb58110e5222984dfa",
                        "test.fa:md5,cd38bf45ddba78787705257b840f6568"
                    ]
                ],
                "purged_haplotigs": [
                    [
                        {
                            "id": "test"
                        },
                        "test.hap.fa:md5,b89d4488c2451413b065b04bb6e08b64"
                    ]
                ],
                "purgedups_bed": [
                    [
                        {
                            "id": "test"
                        },
                        "test.dups.bed:md5,dfb067e8acdfde5c6dbef54a2ce1e247"
                    ]
                ],
                "purgedups_calcuts_cutoffs": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cutoffs:md5,7a0c3b9371c4d1505452f5436f173e1e"
                    ]
                ],
                "purgedups_calcuts_log": [
                    [
                        {
                            "id": "test"
                        },
                        "test.calcuts.log:md5,bb16374f3a2bc6ec445b9ac0b584d05b"
                    ]
                ],
                "purgedups_log": [
                    [
                        {
                            "id": "test"
                        },
                        "test.purge_dups.log:md5,f53e9e39a59a0e7870dd6e7aa047cf22"
                    ]
                ],
                "purgedups_pbcstat_basecov": [
                    [
                        {
                            "id": "test"
                        },
                        "test.PB.base.cov:md5,fe7e7762231da2b809b11eebdfdf8ee3"
                    ]
                ],
                "purgedups_pbcstat_hist": [
                    [
                        {
                            "id": "test"
                        },
                        "test.PB.stat:md5,18e18919c3875b0ef14d557d050543ae"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,6538edf27c63c29d6942c5b0a1c31c3c",
                    "versions.yml:md5,65f957bbaf722197435a5107c6c6b128",
                    "versions.yml:md5,77c810fac8a6f2edaaf050d96f063d99",
                    "versions.yml:md5,a68e25631af50931cf951142081c8bd7",
                    "versions.yml:md5,ab1b6b619f2ad229f2b03e05fad2c409",
                    "versions.yml:md5,ab1b6b619f2ad229f2b03e05fad2c409",
                    "versions.yml:md5,ab1b6b619f2ad229f2b03e05fad2c409",
                    "versions.yml:md5,ab1b6b619f2ad229f2b03e05fad2c409",
                    "versions.yml:md5,ab1b6b619f2ad229f2b03e05fad2c409",
                    "versions.yml:md5,bceaceffe8da91afbb66e351b6a9678d",
                    "versions.yml:md5,c9411452b63f327d2a5b66a82b8fd621",
                    "versions.yml:md5,d9e6066e723f9a40b1e22c33e37f2edf",
                    "versions.yml:md5,f7b0f31fa50eff25352b3aac997dfe71",
                    "versions.yml:md5,f8a504bd74283e317e5e474345c3badc"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-10-31T11:30:38.764174"
    },
    "Undibacterium unclassified - hap1 only - hifi": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.purged.fa:md5,f39e71df41f2d5eb58110e5222984dfa",
                        "test.hap.fa:md5,b89d4488c2451413b065b04bb6e08b64"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test.hap.fa:md5,b89d4488c2451413b065b04bb6e08b64"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test.PB.stat:md5,18e18919c3875b0ef14d557d050543ae"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test"
                        },
                        "test.PB.base.cov:md5,fe7e7762231da2b809b11eebdfdf8ee3"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cutoffs:md5,7a0c3b9371c4d1505452f5436f173e1e"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test"
                        },
                        "test.calcuts.log:md5,bb16374f3a2bc6ec445b9ac0b584d05b"
                    ]
                ],
                "6": [
                    [
                        {
                            "id": "test"
                        },
                        "test.dups.bed:md5,dfb067e8acdfde5c6dbef54a2ce1e247"
                    ]
                ],
                "7": [
                    [
                        {
                            "id": "test"
                        },
                        "test.purge_dups.log:md5,f53e9e39a59a0e7870dd6e7aa047cf22"
                    ]
                ],
                "8": [
                    "versions.yml:md5,6538edf27c63c29d6942c5b0a1c31c3c",
                    "versions.yml:md5,65f957bbaf722197435a5107c6c6b128",
                    "versions.yml:md5,77c810fac8a6f2edaaf050d96f063d99",
                    "versions.yml:md5,a68e25631af50931cf951142081c8bd7",
                    "versions.yml:md5,ab1b6b619f2ad229f2b03e05fad2c409",
                    "versions.yml:md5,ab1b6b619f2ad229f2b03e05fad2c409",
                    "versions.yml:md5,ab1b6b619f2ad229f2b03e05fad2c409",
                    "versions.yml:md5,ab1b6b619f2ad229f2b03e05fad2c409",
                    "versions.yml:md5,ab1b6b619f2ad229f2b03e05fad2c409",
                    "versions.yml:md5,bceaceffe8da91afbb66e351b6a9678d",
                    "versions.yml:md5,c9411452b63f327d2a5b66a82b8fd621",
                    "versions.yml:md5,d9e6066e723f9a40b1e22c33e37f2edf",
                    "versions.yml:md5,f8a504bd74283e317e5e474345c3badc"
                ],
                "purged_assemblies": [
                    [
                        {
                            "id": "test"
                        },
                        "test.purged.fa:md5,f39e71df41f2d5eb58110e5222984dfa",
                        "test.hap.fa:md5,b89d4488c2451413b065b04bb6e08b64"
                    ]
                ],
                "purged_haplotigs": [
                    [
                        {
                            "id": "test"
                        },
                        "test.hap.fa:md5,b89d4488c2451413b065b04bb6e08b64"
                    ]
                ],
                "purgedups_bed": [
                    [
                        {
                            "id": "test"
                        },
                        "test.dups.bed:md5,dfb067e8acdfde5c6dbef54a2ce1e247"
                    ]
                ],
                "purgedups_calcuts_cutoffs": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cutoffs:md5,7a0c3b9371c4d1505452f5436f173e1e"
                    ]
                ],
                "purgedups_calcuts_log": [
                    [
                        {
                            "id": "test"
                        },
                        "test.calcuts.log:md5,bb16374f3a2bc6ec445b9ac0b584d05b"
                    ]
                ],
                "purgedups_log": [
                    [
                        {
                            "id": "test"
                        },
                        "test.purge_dups.log:md5,f53e9e39a59a0e7870dd6e7aa047cf22"
                    ]
                ],
                "purgedups_pbcstat_basecov": [
                    [
                        {
                            "id": "test"
                        },
                        "test.PB.base.cov:md5,fe7e7762231da2b809b11eebdfdf8ee3"
                    ]
                ],
                "purgedups_pbcstat_hist": [
                    [
                        {
                            "id": "test"
                        },
                        "test.PB.stat:md5,18e18919c3875b0ef14d557d050543ae"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,6538edf27c63c29d6942c5b0a1c31c3c",
                    "versions.yml:md5,65f957bbaf722197435a5107c6c6b128",
                    "versions.yml:md5,77c810fac8a6f2edaaf050d96f063d99",
                    "versions.yml:md5,a68e25631af50931cf951142081c8bd7",
                    "versions.yml:md5,ab1b6b619f2ad229f2b03e05fad2c409",
                    "versions.yml:md5,ab1b6b619f2ad229f2b03e05fad2c409",
                    "versions.yml:md5,ab1b6b619f2ad229f2b03e05fad2c409",
                    "versions.yml:md5,ab1b6b619f2ad229f2b03e05fad2c409",
                    "versions.yml:md5,ab1b6b619f2ad229f2b03e05fad2c409",
                    "versions.yml:md5,bceaceffe8da91afbb66e351b6a9678d",
                    "versions.yml:md5,c9411452b63f327d2a5b66a82b8fd621",
                    "versions.yml:md5,d9e6066e723f9a40b1e22c33e37f2edf",
                    "versions.yml:md5,f8a504bd74283e317e5e474345c3badc"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-10-30T11:02:30.336346"
    }
}